clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -1442,13 +1527,46 @@ class SyclKernelBodyCreator
     // TODO: Creates init/finalize sequence and inits special sycl obj
   }
 
+  void CreateExprForStructOrScalar(FieldDecl *FD) {
+    ParmVarDecl *KernelParameter =
+        DeclCreator.getParamVarDeclsForCurrentField()[0];
+    InitializedEntity Entity =
+        InitializedEntity::InitializeMember(FD, &VarEntity);
+    QualType ParamType = KernelParameter->getOriginalType();
+    Expr *DRE = SemaRef.BuildDeclRefExpr(KernelParameter, ParamType, VK_LValue,
+                                         SourceLocation());
+    InitializationKind InitKind =
+        InitializationKind::CreateCopy(SourceLocation(), SourceLocation());
+    InitializationSequence InitSeq(SemaRef, Entity, InitKind, DRE);
+
+    ExprResult MemberInit = InitSeq.Perform(SemaRef, Entity, InitKind, DRE);
+    InitExprs.push_back(MemberInit.get());
+  }
 
   void handleStructType(const FieldDecl *FD, QualType Ty) final {
-    // TODO: a bunch of work doing inits, note this has a little more than
-    // scalar.
+    CreateExprForStructOrScalar(const_cast<FieldDecl *>(FD));
   }
+
   void handleScalarType(const FieldDecl *FD, QualType Ty) final {
-    // TODO: a bunch of work doing inits.
+    CreateExprForStructOrScalar(const_cast<FieldDecl *>(FD));
+  }
+
+  void enterStruct(const CXXRecordDecl *, const FieldDecl *FD) final {
+    CurBase = BuildMemberExpr(CurBase, const_cast<FieldDecl *>(FD));
+    Bases.push_back(CurBase);
+  }
+
+  void leaveStruct(const CXXRecordDecl *, const FieldDecl *FD) final {
+    CurBase = Bases.back();
+    Bases.pop_back();
+  }
+
+  void enterStruct(const CXXRecordDecl *RD, const CXXBaseSpecifier &BS) final {
+    // TODO : do something here?
```

Comment:
These are structs that are 'base' classes, ones picked up via inheritence along the way.